pkg/cmd/openshift-sdn/proxy.go

```diff
@@ -346,3 +352,71 @@ func getNodeIP(client kv1core.CoreV1Interface, hostname string) (net.IP, error)
 
 	return nodeIP, nil
 }
+
+type waitingProxyHandler struct {
+	initialized uint32
+	// WaitChan will be closed when both services and endpoints have
+	// been synced in the proxy
+	waitChan chan<- bool
+
+	serviceChild    pconfig.ServiceHandler
+	serviceSynced   uint32
+	endpointsChild  pconfig.EndpointsHandler
+	endpointsSynced uint32
+}
+
+func newWaitingProxyHandler(serviceChild pconfig.ServiceHandler, endpointsChild pconfig.EndpointsHandler, waitChan chan<- bool) *waitingProxyHandler {
+	return &waitingProxyHandler{
+		serviceChild:   serviceChild,
+		endpointsChild: endpointsChild,
+		waitChan:       waitChan,
+	}
+}
+
+func (wph *waitingProxyHandler) checkInitialized() {
+	if atomic.LoadUint32(&wph.initialized) > 0 {
+		// Don't close our channel twice
+		return
+	}
+	if atomic.LoadUint32(&wph.serviceSynced) > 0 && atomic.LoadUint32(&wph.endpointsSynced) > 0 {
+		klog.V(2).Info("openshift-sdn proxy services and endpoints initialized")
+		atomic.StoreUint32(&wph.initialized, 1)
+		close(wph.waitChan)
+	}
+}
+
+func (wph *waitingProxyHandler) OnServiceAdd(service *v1.Service) {
+	wph.serviceChild.OnServiceAdd(service)
```

Comment:
Do we care if it is already added and/or initialized?

Reply:
> This just sets (writes) the ready file. When we are gracefully going down should we delete the file? What should the state of the file be when we are out of sync (while updating). Do we hold off requests while running sync?

We don't really have a graceful termination handler in the code; that's handled by the container startup script in the CNO YAML. But we already have the patches you did for removing that file in the MCO and container termination hook, which should serve that purpose instead.

Reply:
> Do we care if it is already added and/or initialized?

The underlying proxies themselves also have an 'initialized' check that ensures they don't process anything until OnServiceSynced() is called. So passing these calls along is fine since the proxies are already handling this OK.